x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -116,4 +116,56 @@ export class BulkUploader {
   _onPayload(payload) {
     return sendBulkPayload(this._client, this._interval, payload);
   }
+
+  /*
+   * Bulk stats are transformed into a bulk upload format
+   * Non-legacy transformation is done in CollectorSet.toApiStats
+   */
+  static toBulkUploadFormat(uploadData) {
+    const payload = uploadData
+      .filter(d => Boolean(d) && !isEmpty(d.result))
+      .map(({ result, type }) => [{ index: { _type: type } }, result]);
+    if (payload.length > 0) {
+      const combinedData = BulkUploader.combineStatsLegacy(payload); // arrange the usage data into the stats
+      return flatten(combinedData);
+    }
+  }
+
+  static combineStatsLegacy(payload) {
+    // default the item to [] to allow destructuring
+    const findItem = type => payload.find(item => get(item, '[0].index._type') === type) || [];
+
+    // kibana usage and stats
+    let statsResult;
+    const [ statsHeader, statsPayload ] = findItem(KIBANA_STATS_TYPE);
+    const [ reportingHeader, reportingPayload ] = findItem(KIBANA_REPORTING_TYPE);
+
+    if (statsHeader && statsPayload) {
+      const [ usageHeader, usagePayload ] = findItem(KIBANA_USAGE_TYPE);
+      const kibanaUsage = (usageHeader && usagePayload) ? usagePayload : null;
+      const reportingUsage = (reportingHeader && reportingPayload) ? reportingPayload : null;
+      statsResult = [ statsHeader, statsPayload ];
+      if (kibanaUsage) {
+        set(statsResult, '[1].usage', kibanaUsage);
```

Comment:
Something's not quite right around here. Under `usage` in `.monitoring-kibana-6-*` docs I'm seeing kibana operational stats (like cpu, etc.), not kibana usage (like number of dashboards, etc.).

Reply:
51444d7 - fixed the bug and the issue with the test that silenced the bug

Reply:
@tsullivan I just pulled down the PR again (so I'm on commit SHA 7102f04) and I'm still seeing bad information in the `usage` field.

Concretely, I compared what is indexed into `.monitoring-kibana-6-*` by this PR (that is, without metricbeat in the picture) with what is indexed when running `master`. 

Here's what's under `usage` with this PR:

```js
{
  "kibana":{
    "uuid":"5b2de169-2785-441b-ae8c-186a1936b17d",
    "name":"Shaunaks-MBP-2",
    "index":".kibana",
    "host":"localhost",
    "transport_address":"localhost:5601",
    "version":"7.0.0-alpha1",
    "snapshot":false,
    "status":"green"
  },
  "last_updated":"2018-07-18T15:37:04.711Z",
  "collection_interval_in_millis":5000,
  "process":{
    "memory":{
      "heap":{
        "total_in_bytes":242446336,
        "used_in_bytes":199243704,
        "size_limit":1501560832
      },
      "resident_set_size_in_bytes":342654976
    },
    "event_loop_delay":0.21505308151245117,
    "pid":54887,
    "uptime_in_millis":303398
  },
  "os":{
    "load":{
      "1m":2.3486328125,
      "5m":3.06103515625,
      "15m":2.9541015625
    },
    "memory":{
      "total_in_bytes":17179869184,
      "free_in_bytes":348852224,
      "used_in_bytes":16831016960
    },
    "uptime_in_millis":2241549000
  },
  "response_times":{
    "avg_in_millis":44,
    "max_in_millis":44
  },
  "requests":{
    "total":2,
    "disconnects":0,
    "status_codes":{
      "200":1
    }
  },
  "concurrent_connections":6,
  "xpack":{
    "reporting":{
      "available":true,
      "enabled":true,
      "browser_type":"phantom",
      "_all":0,
      "csv":{
        "available":true,
        "total":0
      },
      "printable_pdf":{
        "available":false,
        "total":0
      },
      "status":{

      },
      "lastDay":{
        "_all":0,
        "csv":{
          "available":true,
          "total":0
        },
        "printable_pdf":{
          "available":false,
          "total":0
        },
        "status":{

        }
      },
      "last7Days":{
        "_all":0,
        "csv":{
          "available":true,
          "total":0
        },
        "printable_pdf":{
          "available":false,
          "total":0
        },
        "status":{

        }
      }
    }
  }
}
```

And here's what's under `usage` with `master`:

```js
{
  "index":".kibana",
  "dashboard":{
    "total":0
  },
  "visualization":{
    "total":0
  },
  "search":{
    "total":0
  },
  "index_pattern":{
    "total":1
  },
  "graph_workspace":{
    "total":0
  },
  "timelion_sheet":{
    "total":0
  },
  "xpack":{
    "reporting":{
      "available":true,
      "enabled":true,
      "browser_type":"phantom",
      "_all":0,
      "csv":{
        "available":true,
        "total":0
      },
      "printable_pdf":{
        "available":false,
        "total":0
      },
      "status":{

      },
      "lastDay":{
        "_all":0,
        "csv":{
          "available":true,
          "total":0
        },
        "printable_pdf":{
          "available":false,
          "total":0
        },
        "status":{

        }
      },
      "last7Days":{
        "_all":0,
        "csv":{
          "available":true,
          "total":0
        },
        "printable_pdf":{
          "available":false,
          "total":0
        },
        "status":{

        }
      }
    }
  }
}
```

### Steps to reproduce
1. Checkout `master`
2. Run Kibana so it indexes a few documents into `.monitoring-kibana-6-*`
3. Grab a `kibana_stats` document
4. Stop Kibana
5. Delete `.monitoring-kibana-6-*` (to start with a clean slate again)
6. Checkout this PR
7. Run Kibana so it indexes a few documents into `.monitoring-kibana-6-*`
8. Grab a `kibana_stats` document
9. Compare the `kibana_stats` document from step 3 with the one from step 8. They should be identical in structure.

Reply:
I'm seeing the same thing ^^^

I wrote a script that automates this for me using [diff](https://github.com/flitbit/diff) and it results in the following fields not appearing in the monitoring documents from this PR:

```
'kibana_stats.usage.index'
'kibana_stats.usage.dashboard'
'kibana_stats.usage.visualization'
'kibana_stats.usage.search'
'kibana_stats.usage.index_pattern'
'kibana_stats.usage.graph_workspace'
'kibana_stats.usage.timelion_sheet'
```

It also tells me that the monitoring documents from this PR add the following fields:
```
'kibana_stats.usage.kibana'
'kibana_stats.usage.last_updated'
'kibana_stats.usage.collection_interval_in_millis'
'kibana_stats.usage.process'
'kibana_stats.usage.os'
'kibana_stats.usage.response_times'
'kibana_stats.usage.requests'
'kibana_stats.usage.concurrent_connections'
```

Reply:
Thanks for catching this! My manual testing steps must be screwy. I'm on this.